cilium/CFP-33716-distributed-packet-tracing.md

```diff
@@ -0,0 +1,586 @@
+
+## Proposal
+
+### Overview
+Offer support for packets containing a user-defined IP option field. For tagged incoming flows, handle events for this flow specially; emit datapath debugging events for these packets, include the ID in the event message, and implement hubble filtering by the ID.
+
+### Section 1: Packet Tagging 
+### Packet Tagging Responsibility
+
+This feature does not implement the packet tagging (see Deferred Milestone 1: Tag Packets). Instead, it is the client's responsibility to tag packets via IP option embedding. Therefore, the key assumption for this feature is that tagged packets arrive at ingress points with an `ip_trace_id` already embedded in the IP options.
+
+#### Reasons for Not Implementing Automatic Packet Tagging:
+
+1. **Avoid MTU Concerns:**
+   - Automatically tagging packets can cause them to exceed the MTU, leading to fragmentation and performance issues.
+
+2. **Prevent Performance Loss from GRO:**
+   - GRO engine does not support packets with IP options.
+
+3. **Avoid Packet Loss:**
+   - Some network devices and paths may not support IP options, causing tagged packets in the path to be dropped.
+   - Targets that don't support IP Options:
+      - L4 ELB
+   - Targets that support IP Options:
+      - L4 ILB
+      - Services
+      - Pod-to-pod (including across VPC)
+
+4. **IP Checksum Recalculation:**
+   - When IP options are added or modified, the IP checksum must be recalculated, adding additional processing overhead and complexity.
+
+By making packet tagging the client's responsibility, we ensure that Cilium remains unaffected by these concerns, leaving it to the clients to incorporate these key considerations into their use of this feature and decide how best to utilize it to suit their specific needs.
+
+### Section 2: Feature Enablement
+`--enable-ip-option-tracing=<type>` to define the IP option type used to embed the trace ID value
+- `<type>` corresponds to the IP option value ([https://www.iana.org/assignments/ip-parameters/ip-parameters.xhtml](https://www.iana.org/assignments/ip-parameters/ip-parameters.xhtml))
+- By default, the feature is disabled (`<type>=0`).
+- `<type>=0` disables this feature.
+
+```go
+EnableIPOptionTracing = "enable-ip-option-tracing"
+```
+
+```go
+if option.Config.EnableIPOptionTracing > 0 {
+    cDefinesMap["ENABLE_PACKET_IP_TRACING"] = 
+    strconv.Itoa(option.Config.EnableIPOptionTracing)
+}
+```
+
+### Section 3: Parsing
+
+Parsing at the ingress point of the TC hook allows for early detection and tracing of a flow.
+
+```c
+#ifdef ENABLE_PACKET_IP_TRACING
+    check_and_store_ip_trace_id(ctx, ENABLE_PACKET_IP_TRACING);
+    // Do parsing and storing of ip_trace_id
+#endif
+```
+```c
+// Function to parse trace ID and store it if valid
+static __always_inline void check_and_store_ip_trace_id(struct __ctx_buff *ctx, __u8 ip_opt_type_value) {
```

Comment:
(minor) Seems like the #define is both a CPP define and passed as a param into the function. Any particular reason why it is doing double duty?

Reply:
When a user sets the feature-gate to enable this feature, they do so by setting the specific IP option where the packets `ip_trace_id`s are embedded. So, we want to pass this into the parsing function in order to know where to read these `ip_trace_id`s from. 

The implementation specifics can be viewed [here](https://github.com/cilium/cilium/pull/33957/files) in the `pkg/datapath/linux/config/config.go` file if you're interested in more context.